File: web/src/features/graph/normalization.ts
import { isCausalGraphSpec, type ComponentSpec, type GraphSpec, type TapSpec } from '@/types/graph';
import type { StudioTaskBindingSpec, StudioWorkspaceSpec } from '@/types/workspace';
import { taskBindingId } from '@/features/scenario/taskBindings';
import { normalizeDynamicPorts } from '@/features/graph/dynamicPorts';

function createTapId() {
  if (typeof crypto !== 'undefined' && 'randomUUID' in crypto) {
    return crypto.randomUUID();
  }
  return `tap-${Date.now()}-${Math.floor(Math.random() * 1_000_000)}`;
}

function migrateLegacyTaps(graph: GraphSpec): TapSpec[] {
  const taps: TapSpec[] = graph.taps ? [...graph.taps] : [];
  const usedIds = new Set(taps.map((tap) => tap.id));

  const addTap = (tap: TapSpec) => {
    if (usedIds.has(tap.id)) {
      tap = { ...tap, id: createTapId() };
    }
    usedIds.add(tap.id);
    taps.push(tap);
  };

  if (graph.barnacles) {
    for (const [nodeId, barnacles] of Object.entries(graph.barnacles)) {
      for (const barnacle of barnacles) {
        const paths: Record<string, string> = {};
        const usedNames = new Set<string>();
        for (const path of barnacle.read_paths ?? []) {
          const base = path.split('.').slice(-1)[0] || 'value';
          let name = base;
          let idx = 2;
          while (usedNames.has(name)) {
            name = `${base}_${idx}`;
            idx += 1;
          }
          usedNames.add(name);
          paths[name] = path;
        }
        const transform =
          barnacle.kind === 'intervention'
            ? {
                type: barnacle.label || 'intervention',
                params: {
                  read_paths: barnacle.read_paths ?? [],
                  write_paths: barnacle.write_paths ?? [],
                  transform: barnacle.transform ?? '',
                },
              }
            : undefined;
        addTap({
          id: barnacle.id,
          type: barnacle.kind,
          position: { afterNode: nodeId },
          paths,
          transform,
        });
      }
    }
  }

  if (graph.user_ports) {
    for (const [nodeId, ports] of Object.entries(graph.user_ports)) {
      const paths: Record<string, string> = {};
      for (const port of ports.outputs ?? []) {
        paths[port] = port;
      }
      for (const port of ports.inputs ?? []) {
        if (!(port in paths)) {
          paths[port] = port;
        }
      }
      if (Object.keys(paths).length > 0) {
        addTap({
          id: createTapId(),
          type: 'probe',
          position: { afterNode: nodeId },
          paths,
        });
      }
    }
  }

  return taps;
}

export function normalizeGraphAuthoringTypes(graph: GraphSpec): GraphSpec {
  const renamePort = (
    nodeId: string,
    port: string,
    spec: ComponentSpec
  ) => {
    if (spec.type === 'Network' && port === 'target') {
      return 'input';
    }
    return port;
  };

  const nodes = Object.fromEntries(
    Object.entries(graph.nodes).map(([id, spec]) => {
      let nextType = spec.type;
      if (nextType === 'FeedbackChannel') nextType = 'Channel';
      if (nextType === 'PenzaiSubgraph') nextType = 'PenzaiAdapter';
      const nextParams = { ...spec.params };
      const nextSpec: ComponentSpec = {
        ...spec,
        type: nextType,
        params: nextParams,
      };
      return [id, nextSpec];
    })
  );
  const wires = graph.wires.map((wire) => {
    const sourceSpec = nodes[wire.source_node];
    const targetSpec = nodes[wire.target_node];
    const normalizedWire = {
      ...wire,
      source_port: sourceSpec ? renamePort(wire.source_node, wire.source_port, sourceSpec) : wire.source_port,
      target_port: targetSpec ? renamePort(wire.target_node, wire.target_port, targetSpec) : wire.target_port,
      temporality: wire.temporality,
      recurrent_initializer: wire.recurrent_initializer,
    };
    return normalizedWire;
  });
  const input_bindings = Object.fromEntries(
    Object.entries(graph.input_bindings).map(([name, binding]) => {
      const [nodeId, port] = binding;
      const spec = nodes[nodeId];
      const nextPort = spec ? renamePort(nodeId, port, spec) : port;
      return [name === 'target' ? 'input' : name, [nodeId, nextPort] as [string, string]];
    })
  );
  const input_ports = graph.input_ports.map((port) => (port === 'target' ? 'input' : port));
  const subgraphs =
    graph.subgraphs
      ? Object.fromEntries(
          Object.entries(graph.subgraphs).map(([id, subgraph]) => [
            id,
            isCausalGraphSpec(subgraph) ? normalizeGraphAuthoringTypes(subgraph) : subgraph,
          ])
        )
      : undefined;
  const taps =
    graph.taps || graph.barnacles || graph.user_ports ? migrateLegacyTaps(graph) : undefined;
  const normalized: GraphSpec = {
    ...graph,
    nodes,
    wires,
    input_ports,
    input_bindings,
  };
  if (taps) normalized.taps = taps;
  if (subgraphs) normalized.subgraphs = subgraphs;
  if (graph.barnacles) normalized.barnacles = undefined;
  if (graph.user_ports) normalized.user_ports = undefined;
  return normalized;
}

export function normalizeGraphForStudioAuthoring(
  graph: GraphSpec,
  taskBindingSpec?: StudioTaskBindingSpec | null
): GraphSpec {
  return normalizeDynamicPorts(normalizeGraphAuthoringTypes(graph), taskBindingSpec);
}

export function normalizeTaskBindingSpecForStudioAuthoring(
  taskBindingSpec: StudioTaskBindingSpec | null | undefined,
  graph: GraphSpec
): StudioTaskBindingSpec | null | undefined {
  if (!taskBindingSpec) return taskBindingSpec;
  let changed = false;
  const bindings = taskBindingSpec.bindings.map((binding) => {
    if (
      binding.target_node_id === 'mux' &&
      !graph.nodes.mux &&
      graph.nodes.input_mux &&
      graph.nodes.input_mux.type === 'Mux'
    ) {
      changed = true;
      return {
        ...binding,
        id: taskBindingId(
          binding.source_data_id,
          'input_mux',
          binding.target_port,
          binding.target_graph_path
        ),
        target_node_id: 'input_mux',
      };
    }
    const target = graph.nodes[binding.target_node_id];
    if (target?.type !== 'Network' || binding.target_port !== 'target') {
      return binding;
    }
    changed = true;
    return {
      ...binding,
      id: taskBindingId(
        binding.source_data_id,
        binding.target_node_id,
        'input',
        binding.target_graph_path
      ),
      target_port: 'input',
    };
  });
  return changed ? { ...taskBindingSpec, bindings } : taskBindingSpec;
}

export function normalizeWorkspaceGraphsForStudioAuthoring(
  workspace: StudioWorkspaceSpec | null
): StudioWorkspaceSpec | null {
  if (!workspace) return workspace;
  let changed = false;
  const scenarios = Object.fromEntries(
    Object.entries(workspace.scenarios).map(([scenarioId, scenario]) => {
      if (!scenario.graph) return [scenarioId, scenario];
      const graph = normalizeGraphForStudioAuthoring(
        scenario.graph,
        scenario.task_binding_spec
      );
      const taskBindingSpec = normalizeTaskBindingSpecForStudioAuthoring(
        scenario.task_binding_spec,
        graph
      );
      if (graph === scenario.graph && taskBindingSpec === scenario.task_binding_spec) {
        return [scenarioId, scenario];
      }
      changed = true;
      return [scenarioId, { ...scenario, graph, task_binding_spec: taskBindingSpec }];
    })
  );
  return changed ? { ...workspace, scenarios } : workspace;
}

File: feedbax/contracts/graphs/normalization.py
"""Studio authoring normalization for persisted graph specs."""

from __future__ import annotations

from typing import Optional

from feedbax.contracts.acausal_interface import normalize_acausal_interfaces_for_graph
from feedbax.contracts.graph import (
    ComponentSpec,
    GraphProject,
    GraphSpec,
    StudioTaskBindingSpec,
    StudioWorkspaceSpec,
    WireSpec,
)


def _authoring_component_type(component_type: str) -> str:
    if component_type == "FeedbackChannel":
        return "Channel"
    if component_type == "PenzaiSubgraph":
        return "PenzaiAdapter"
    return component_type


def normalize_graph_for_studio_authoring(graph: GraphSpec) -> GraphSpec:
    """Normalize runtime/persisted component names to Studio authoring names."""

    graph = normalize_acausal_interfaces_for_graph(graph)
    nodes: dict[str, ComponentSpec] = {}
    for node_id, node_spec in graph.nodes.items():
        next_type = _authoring_component_type(node_spec.type)
        params = dict(node_spec.params)
        input_ports = list(node_spec.input_ports)
        output_ports = list(node_spec.output_ports)
        nodes[node_id] = node_spec.model_copy(
            update={
                "type": next_type,
                "params": params,
                "input_ports": input_ports,
                "output_ports": output_ports,
            }
        )

    def rename_port(node_name: str, port: str) -> str:
        node = nodes.get(node_name)
        if node and node.type == "Network" and port == "target":
            return "input"
        return port

    wires: list[WireSpec] = []
    for wire in graph.wires:
        normalized_wire = WireSpec(
            source_node=wire.source_node,
            source_port=rename_port(wire.source_node, wire.source_port),
            target_node=wire.target_node,
            target_port=rename_port(wire.target_node, wire.target_port),
            temporality=wire.temporality,
            recurrent_initializer=wire.recurrent_initializer,
        )
        wires.append(normalized_wire)
    input_bindings = {
        ("input" if name == "target" else name): (
            node,
            rename_port(node, port),
        )
        for name, (node, port) in graph.input_bindings.items()
    }
    subgraphs = None
    if graph.subgraphs:
        subgraphs = {}
        for node_id, subgraph in graph.subgraphs.items():
            if isinstance(subgraph, GraphSpec):
                subgraphs[node_id] = normalize_graph_for_studio_authoring(subgraph)
            else:
                subgraphs[node_id] = subgraph
    return graph.model_copy(
        update={
            "nodes": nodes,
            "wires": wires,
            "input_ports": ["input" if port == "target" else port for port in graph.input_ports],
            "input_bindings": input_bindings,
            "subgraphs": subgraphs,
        }
    )


def normalize_task_binding_spec_for_studio_authoring(
    task_binding_spec: Optional[StudioTaskBindingSpec],
    graph: GraphSpec,
) -> Optional[StudioTaskBindingSpec]:
    if task_binding_spec is None:
        return None
    changed = False
    bindings = []
    for binding in task_binding_spec.bindings:
        target = graph.nodes.get(binding.target_node_id)
        if target is not None and target.type == "Network" and binding.target_port == "target":
            binding = binding.model_copy(
                update={
                    "id": (f"task:{binding.source_data_id}->{binding.target_node_id}:input"),
                    "target_port": "input",
                }
            )
            changed = True
        bindings.append(binding)
    return (
        task_binding_spec.model_copy(update={"bindings": bindings})
        if changed
        else task_binding_spec
    )


def normalize_workspace_for_studio_authoring(
    workspace: Optional[StudioWorkspaceSpec],
) -> Optional[StudioWorkspaceSpec]:
    if workspace is None:
        return None
    changed = False
    scenarios = dict(workspace.scenarios)
    for scenario_id, scenario in workspace.scenarios.items():
        if scenario.graph is None:
            continue
        graph = normalize_graph_for_studio_authoring(scenario.graph)
        task_binding_spec = normalize_task_binding_spec_for_studio_authoring(
            scenario.task_binding_spec,
            graph,
        )
        if graph is scenario.graph and task_binding_spec is scenario.task_binding_spec:
            continue
        scenarios[scenario_id] = scenario.model_copy(
            update={"graph": graph, "task_binding_spec": task_binding_spec}
        )
        changed = True
    return workspace.model_copy(update={"scenarios": scenarios}) if changed else workspace


def normalize_project_for_studio_authoring(project: GraphProject) -> GraphProject:
    graph = normalize_graph_for_studio_authoring(project.graph)
    workspace = normalize_workspace_for_studio_authoring(project.workspace)
    if graph is project.graph and workspace is project.workspace:
        return project
    return project.model_copy(update={"graph": graph, "workspace": workspace})
